src/sentry/sentry_metrics/consumers/indexer/batch.py

```diff
@@ -98,94 +99,136 @@ def _extract_namespace(self, headers: Headers) -> Optional[str]:
+            broker_meta = BrokerMeta(msg.value.partition, msg.value.offset)
 
             if (namespace := self._extract_namespace(msg.payload.headers)) in options.get(
                 "sentry-metrics.indexer.disabled-namespaces"
             ):
                 assert namespace
-                self.skipped_offsets.add(partition_offset)
-                disabled_msgs_cnt[namespace] += 1
-                continue
-            try:
-                parsed_payload: ParsedMessage = json.loads(
-                    msg.payload.value.decode("utf-8"), use_rapid_json=True
-                )
-            except rapidjson.JSONDecodeError:
-                self.skipped_offsets.add(partition_offset)
-                logger.error(
-                    "process_messages.invalid_json",
-                    extra={"payload_value": str(msg.payload.value)},
-                    exc_info=True,
-                )
+                skipped_msgs_cnt[namespace] += 1
+                self.filtered_msg_meta.add(broker_meta)
                 continue
-            try:
-                if self.__input_codec:
-                    self.__input_codec.validate(parsed_payload)
-            except ValidationError:
-                if settings.SENTRY_METRICS_INDEXER_RAISE_VALIDATION_ERRORS:
-                    raise
-
-                # For now while this is still experimental, those errors are
-                # not supposed to be fatal.
-                logger.warning(
-                    "process_messages.invalid_schema",
-                    extra={"payload_value": str(msg.payload.value)},
-                    exc_info=True,
-                )
 
             try:
-                parsed_payload["use_case_id"] = use_case_id = extract_use_case_id(
-                    parsed_payload["name"]
-                )
-            except ValidationError:
-                self.skipped_offsets.add(partition_offset)
-                logger.error(
-                    "process_messages.invalid_metric_resource_identifier",
-                    extra={"payload_value": str(msg.payload.value)},
-                    exc_info=True,
-                )
-                continue
-
-            self.__message_count[use_case_id] += 1
-            self.__message_size_max[use_case_id] = max(
-                len(msg.payload.value), self.__message_size_max[use_case_id]
-            )
-            self.__message_size_sum[use_case_id] += len(msg.payload.value)
-
-            # Ensure that the parsed_payload can be cast back to to
-            # IngestMetric. If there are any schema changes, this check would
-            # fail and ParsedMessage needs to be adjusted to be a superset of
-            # IngestMetric again.
-            _: IngestMetric = parsed_payload
-
-            self.parsed_payloads_by_offset[partition_offset] = parsed_payload
+                parsed_payload = self._extract_message(msg)
+                self._validate_message(parsed_payload)
```

Comment:
Both _extract_message and _validate_message seem to be performing some form of validations. I would keep a single method to cover the validations. That would ensure high cohesion in the singular function responsible for handling data quality problems.

Reply:
Thanks, here is my original line of thinking for separating it into those two methods:

`_extract_message` deserializes a `KafkaPayload` (which is pretty much the raw bytes) into a `ParsedMessage` data structure in python. While `_validate_message` actually semantically verifies that the `ParsedMessage` contents are valid.

I would argue that those two steps should be separate, despite that `_extract_message` performs some checks, they are just there to make sure the **types** are correct, here are the 3 things it checks:
1. The raw bytes can be decoded into a dictionary
2. The dictionary contains all the fields the `IngestMetric` type dict (which is a superclass of `ParsedMessage`
3. Make sure that we can deserialize the MetricID into a `UseCaseID`, which is a requirement of `ParsedMessage`

Meanwhile, `_validate_message` actually checks the content of the types, such as the metric name is not too long and the tags are valid.

Please let me know if this makes sense

Reply:
> _extract_message deserializes a KafkaPayload (which is pretty much the raw bytes) into a ParsedMessage data structure in python. While _validate_message actually semantically verifies that the ParsedMessage contents are valid.

We can have 2 separate methods then. A minor improvement I would suggest then is to document some of these methods better. The IndexerBatch is so core to the metrics indexer and could use some better documentation. No need to add documentation to the whole class. But when you are touching a method, feel free to add some method documentation.

Reply:
Sounds good, I added some pretty comprehensive comments in [processing.py](https://github.com/getsentry/sentry/pull/58675/files#diff-79c172ff490e45a4616dd358296c820ad37050f525d14aa874f2699b8c06953cR85-R102) and [batch.py](https://github.com/getsentry/sentry/pull/58675/files#diff-7697ef4fd4cf439ed7ad41b7d98c00ff7d24cde30953e12d6bcb877303bb942dR98-R109)